### ISSUE TITLE
InitGeolocation: run at every boot / incorrect info via VPN

### ISSUE THREAD

**Issue by Azureit** [openatv/enigma2#1563]
What is the need to run this at every boot when we already manually configured language/timezone?
And this will produce incorrect info if I VPN my internet.

**Comment by IanSav** [comment]
The code runs to initialise the geolocation data.  The data is available to *any* Enigma2 code that wants to use it.

The code has been updated to no longer return *any* data if a VPN is detected.

**Comment by Azureit** [comment]
@IanSav Why not cache the data and reuse, next boot.

**Comment by IanSav** [comment]
Some of the data captured may change.  The only way to validate the cache would be to do another lookup.  ;)

**Comment by Tony-il-Capo** [comment]
> Why not cache the data and reuse next boot.

I manifested already [various concerns about this](https://github.com/openatv/enigma2/commit/b8d1158d558320c2fec9eac90f1b8b57c21c70fd#commitcomment-37733643)

I also wonder if a discussion about this took place before, I hope that geolocation implementation is not a single contributor point-of-view.

**Comment by Azureit** [comment]
@Tony-il-Capo @IanSav
Running a tracker at every boot is privacy concern and in my case will be blocked by captive portal & firewall. To make things worst it's http.

For me the code is good at first boot to assist the configuration, and to cache the results for future use cases. Then the only thing that should change is the IP, but for what!?

The only use of this would be if I was doing around the world with my enigma2 box,  then I will like it to update my timezone and possible the weather app location :) but then again not the language.

**Comment by IanSav** [comment]
Here is the [Privacy Policy](https://ip-api.com/docs/legal) for the site I am using in the code.  I see no malicious intent or any data collection.

This is *NOT* a tracker.  It is a data lookup service.  The only tracking is to hold an IP address, in RAM, for 1 minute to ensure that the service is not being abused.

For the moment, once the user sets a time zone this code makes no further changes.  I am expecting to see requests from users who travel with their receivers to want the time zone to update as they move from place to place.  This capability can easily be added with this new infrastructure.

Exact location accuracy is not particularly important for things like time zones.  The information can always be corrected by the user in the Enigma2 user interface.  Users can also send updates and corrections to the geolocation service if they wish.

**Comment by Tony-il-Capo** [comment]
> I see no malicious intent or any data collection.

Even **if** the server doesn't log any data, it still a privacy issue that you refuse to see and keep ignoring that this creates unwanted mistrust in this project.
ip-api.com quote

> Enrich the data about your users with the city, region, zip, country, continent, timezone, latitude and longitude fields
Find out if users are using anonymous proxies or VPN services
Identify the organization, ISP, AS, and if they are using a mobile network

 and then sends all that openly over internet.
It really sounds like tracking to me.

This should have been implemented in a total different way, **_first create means of user to give consent or not_**, starting from the very first fresh E2 (setup) boot, and only then use this "service".

**Comment by IanSav** [comment]
Please specify exactly what privacy issues are being caused in this case.  In what way is the privacy being disclosed?  What private data is being provided to this site?

Tracking implies that the data is captured and used for some other purpose.  I see *no* evidence of that here.  This is simply an enquiry to deliver publicly available core Internet information back to the initiating request.  No data is collected from the user and there is no evidence that any data is sent elsewhere.

Please detail your proposed solution to achieve the required result within the Enigma2 context.

Given your concerns about privacy I assume that you don't use the Internet at all.  ;)  I suspect that your posts here attract far more tracking and data retention using information that you have provided than anything being done in the geolocation lookup.  Remember that all you are providing is your current WAN IP address.  By definition this information is provided for *every* connection you make on the Internet!

**Comment by Azureit** [comment]
My concerns are: 
- http request can be packet sniffed;
- I don't trust this companies data retention and usage of the data;
- fingerprinting do to "run at every boot", and http request query and headers. 
- I don't want ads because I have a enigma2 box and timed when I turn on the box.

Please explain me why this should run at every boot when I have all configured timezone/language... ?
This geolocation lookup should be initialized when needed, not at every boot.

**Comment by vladc** [comment]
I'm from the staff that operates ip-api.com. a service that launched in 2012. We're a real company in EU, a region with many privacy laws and GDPR.
Our privacy policy is accurate. We do not have any logging for our endpoint at ip-api.com, we answer more than 1.5B requests _every_ day and we have no analytical use for the IP addresses being looked up. As IanSav mentioned, we hold the calling IP address in RAM, for 1 minute to ensure that the service is not being abused (rate limiting), that's all.

**Comment by Azureit** [comment]
@vladc Thanks for taking the time, most appreciated. 
http is unsafe for this, and you can agree with me, that's why https is not for Free API, and the issue is bigger because most of the Free API servers are outside EU.

**Comment by vladc** [comment]
@Azureit while SSL is generally always better, I don't think we can call http unsafe for this. If the code that parses our response has no vulnerability, I'm unsure what impact, if any, a MITM attack can have. Your IP address will always be made public, SSL or otherwise, and the response we return is available for anyone as long as they have your IP address.

**Comment by Tony-il-Capo** [comment]
@IanSav why can't you understand a simple thing, give power to user enable/disable this and the problem is solved. Both sides happy. That way you cover all situations.

E2 has worked for years without this, is not like this is mandatory.

You seem to always trying 'to reinvent the wheel' here and then, defend a square one at all costs. It's not the first time you push very hard for changes in E2 code ignoring even senior contributors opinions. This is a community, we all should be open to suggestions and different ideas, and then try to find a right balance.

As this thing go, I'm out, do whatever you want, I have like ten ways to make sure my boxes never use that 'service', but I feel really bad for the users that are unaware of what is going on.

Bye.

**Comment by IanSav** [comment]
@Tony-il-Capo I am sorry you feel this way.  I am forced to conclude that you are scare mongering and/or you don't understand how networking connections work.  All connections to the Internet work by machines making initial *unsecured* connections to set up the secured connection paths. Each and every one of those unsecured connections have your IP address and probably much more other information.  In this case *only* your current WAN IP address is used.  You should also note that I do *not* provide the IP address it is derived from the connection.

In this case there is a one to one connection from the local / home network to ip-api.com to get the information.  There are no redirections or other sites involved.  The data interchanged does not have any advertising or virus payload.  Even if there was advertising or a virus sent it is not read, processed or used!

If you don't like my contributions to Enigma2 then please raise your issues with Captain and get him to speak to me.  It would be significantly easier for me to abandon OpenATV and only provide my fixes and improvements to the other images.  I am simply including OpenATV in my developments in response to requests and discussions with Caption.

**Comment by atvcaptain** [comment]
@ all i see no Privacy break, only the ip send to the server was used  no other information was used, every connection send the ip to the server, and a secure connection not needed too the ip is not a secret part. location infos are allways presend in most of all other systems, this help user get currect weahter infos, or pc browser show infos for your langauge all use the public ip.

**Comment by IanSav** [comment]
@Azureit:

> * http request can be packet sniffed;

There is no personal data in the request.  It is a simple URL with nothing other than the requested data flag.  There is nothing of value or information to be sniffed.

> * I don't trust this companies data retention and usage of the data;

As you don't provide any data I don't see how this is a concern.  The company is simply returning accumulated public information derived from your ISP.

> * fingerprinting do to "run at every boot", and http request query and headers.

I don't understand this concern.

> * I don't want ads because I have a enigma2 box and timed when I turn on the box.

There are no ads or any other payload other than the requested json data.  Any attempt to inject any ads, viruses or any other other data will cause the geolocation lookup to fail and return no data.

> Please explain me why this should run at every boot when I have all configured timezone/language... ?

This point has already been covered.

> This geolocation lookup should be initialized when needed, not at every boot.

The information provided by the lookup *could* change from boot to boot.

I have thought about this feature for some time.  I did not add it on a whim.  It is there to assist users with their system configuration.  As all the information is triggered from an open and public IP address and the data returned is also from freely available information I considered this to not create a security vulnerability.

The fact that I am continuing to engage in these discussions should be an indication that I take security concerns seriously and stand ready to address any real problems with the code.

**Comment by atvcaptain** [comment]
@IanSav  but we have a major issue

box boot and show wrong clock

after reboot clock was ok but after some time the clock change 1 hour

![Zeit 1](https://user-images.githubusercontent.com/974786/76928732-ac401580-68e2-11ea-84d8-10690c934af0.jpg)

![Zeit 2](https://user-images.githubusercontent.com/974786/76928745-b06c3300-68e2-11ea-9996-451eaa47d4b3.jpg)

we send log infos private

and we dont must set the timezone on every boot, normal only in the startwizard, user get from the wizard  matching defaults for his location, like timezone, ui language, weather location , user can used it or used his own settings

**Comment by atvcaptain** [comment]
at the moment all deep recording are broken we need soon a solution

**Comment by nickersk** [comment]
tz is automatically setting to UTC despite setting it manually ...

**Comment by IanSav** [comment]
There seems to be an error and disparity between the "/etc/localtime" and Enigma2 timezone settings.
If the user goes into the Enigma2 Timezone settings UI and selects the correct timezone the issue should go away.


Are the any users having the issue for old timezones other than UTC?

**Comment by nickersk** [comment]
so, just now i manually set to europe slovakia, and restart gui. after restart in menu there is germany berlin and in etc link is /etc/localtime -> /usr/share/zoneinfo/Europe/Bratislava (which is correct)

**Comment by nickersk** [comment]
and now after some time it jumped to slovakia

**Comment by IanSav** [comment]
I am also in Skype with Captain trying to understand the issue.

**Comment by nickersk** [comment]
after full reboot lrwxrwxrwx    1 root     root            23 Mar 18 06:54 /etc/localtime -> /usr/share/zoneinfo/UTC

**Comment by IanSav** [comment]
A patch has been comitted that should address the time zone issues while we try to find and fix the cause.

**Comment by Tony-il-Capo** [comment]
> @Tony-il-Capo I am sorry you feel this way. I am forced to conclude that you are scare mongering and/or you don't understand how networking connections work. All connections to the Internet work by machines making initial _unsecured_ connections to set up the secured connection paths. Each and every one of those unsecured connections have your IP address and probably much more other information. In this case _only_ your current WAN IP address is used. You should also note that I do _not_ provide the IP address it is derived from the connection.

:rofl: 
Are you serious?!! Now you threat end users as very stupid people? I really can't believe that you now tried to lecture me on how internet works.
Unbelievable.

It's like a stone wall, this all conversation could be resumed to:

- IanSav: I'm gonna completely change this code and hope for the best
- end users: there's issues and concerns...
- IanSav: you're wrong...
- end users: but there are serious problems... we suggest make it optional.
- IanSav: I will teach you how internet works. I know everything!
- IanSav: you're wrong again!

Serious, a little more modesty and understanding would be appreciated.

**FYI** there are VPNs/proxy detected as non proxy (direct connection)...so geolocation data is used but is wrong
eg  some VPN/TOR exit nodes popular for accessing georestricted services.

**Comment by IanSav** [comment]
@Tony-il-Capo I don't believe the rubbish you have been spouting.  Your arguments defy how Internet connections work.  It is demonstrating some serious issues with your scope of knowledge.  When I asked you to provide constructive suggestions on how to realistically address your concerns you simply respond with insults.  A totally unproductive logic that will not convince me or anyone else to change things.

Did you raise your concerns with OpenATV team leaders?  I suspect not as I am working with them on a bug in OpenATV that is causing some of the time zone issues.  They are very keen and happy to have the geolocation code.

When someone gives me log files that demonstrate real issues I will address them.  Some serious debugging is currently happening with Captain and NickerSK trying to find the *real* cause of the time shift issues on OpenATV.  A work around has already been merged and fixes the issue but we all want to find the underlying cause of the bug.  It does not appear to be coming from any of my code.  Something is changing /etc/localtime and resetting it to UTC without any user intervention.

I don't know why am am telling you the real problem as you are already convinced that I don't know anything and am just wasting everyone's time.  How much code have you contributed to Enigma2?

By the way, attempting to bully me is not going to change my mind that the issues about which you complain are non issues.  When you provide some facts, logic and/or detail to your complaint then you may get a more favourable hearing.

**Comment by Tony-il-Capo** [comment]
@IanSav Insults?? Where? Telling you about a different perspective is insulting you?
I'm not sure **what you want me** to do or say?? You now just keep saying I know nothing, ok then! 
I already told you many times that geolocation should be optional. Should be me implementing it? 
Well, I'm not a (python) programmer, I just do programming for personal use only.

I think was clear already that here in my community there are several E2 users that have no background knowledge..., I set up most of them in the past using OpenATV and they learned essential/basic GUI usage. Now it seems I've made the mistake of updating most of them.
I'm here passing the issues I here from them,  do you want me asks those people to debug? Really???
You probably think that every E2 user is here on github!

Believe it or not, for me personally, as I told you before I don't care anymore.
I came here (again) believing it was the right thing to do, with best intentions, but I cannot reason with closed mind self-centered people that only care and see what's inside their bubble.
I now unsubscribed notifications to this, I'm really tired of it.

**Comment by Azureit** [comment]
@atvcaptain 
> @ all i see no Privacy break, only the ip send to the server was used no other information was used, every connection send the ip to the server, and a secure connection not needed too the ip is not a secret part. location infos are allways presend in most of all other systems, this help user get currect weahter infos, or pc browser show infos for your langauge all use the public ip.

This geolocation run at every boot, can be used to fingerprint the user home IP with the use of enigma2 box and the time the user is at home and boot the box. With that info a ads provider, running at a site the user visits, know that user home IP should have ads about, satellite equipment, cable companies, iptv packages...
What's the use of initializing geolocation info at every boot when we have manually configured all? This should be initialized when and if needed. 
At least convince @IanSav the #unmoveblerock to use a far better opensource alternative https://geoip.fedoraproject.org/city The source to run this service is at https://github.com/fedora-infra/geoip-city-wsgi , even openatv could host this service.
I'm advocating for better privacy in this opensource project not for me, that I'm aware of the problems, but to everyone else. I can't imagine for example fedora doing this at every boot, just for the lulz.

**Comment by IanSav** [comment]
@Azureit I tried the site you suggested and it returned the wrong location for me.

I don't understand how a connection on the Internet creates a "fingerprint".  All connections on the Internet share IP addresses.  That can't be stopped or hidden.  It is the fundamental way that the Internet and Ethernet works.

As for the geolocation call, I don't provide *any* information other than the data I want returned.  I don't accept any data back from the geolocation lookup except for a valid json answer.  Anything else returned is just thrown away.  (I have never seen anything else returned.)

Neither you nor Tony have provided *any* specific details about how your privacy is being compromised.  What private data is being violated and what information is being leaked from your connection or your receiver?  Your public / WAN IP address is very much *not* private.

**Comment by Azureit** [comment]
@IanSav 
> I tried the site you suggested and it returned the wrong location for me.

Report that to project, for me it gives a accurate position than ip-api.com.

> I don't understand how a connection on the Internet creates a "fingerprint"

The URI of the request http://ip-api.com/json/?fields=33288191, that "?fields=33288191" is one way to fingerprint, another way can be default urlopen useragent, not every application uses python urlopen... combine that with the WAN IP of the user and you have a fingerprint.

You're still unable to have a valid reason why this should run at every boot ignoring my default settings even if I don't need this, don't want this.

**Comment by atvcaptain** [comment]
you can add to /etc/hosts 
127.0.0.1 ip-api.com and api stop working

**Comment by Azureit** [comment]
@atvcaptain Can you please add a Privacy ON/OFF button in the interface, to do that. Thanks

**Comment by IanSav** [comment]
@Azureit if it were possible to do that I would have done it as part of the initial code implementation.

The mytest.py initialisation script for Enigma2 is quite disorganised and chaotic.  There is no UI and the config engine is not available until considerably far down the code.  I intend to address this at some point in the future but for now what you ask is not yet possible.  The suggestion by @atvcaptain would work for you.

The data "?fields=33288191" in the query is not a fingerprint.  It is simply a bit map that describes the data I want returned  from the geolocation service.  There is absolutely *nothing* about you in that data.  *ALL* Enigma2 users on all images are using *exactly* the same lookup request.

For your information here is how that number is determined: [https://ip-api.com/docs/api:json](https://ip-api.com/docs/api:json).  Feel free to explore the number generator on that page to see how the 33288191 number was generated.  Again there is *nothing* personal in this number and it is *not* a fingerprint that makes your query and more identifiable that *any* other user of that site.

**Comment by Azureit** [comment]
@IanSav 
> The data "?fields=33288191" in the query is not a fingerprint. It is simply a bit map that describes the data I want returned from the geolocation service. There is absolutely nothing about you in that data. 

I know "?fields=33288191" is static not about me, it fingerprint every enigma2 with your code, to associate with me they have my WAN IP.

___________________________________

**I did not test this**, but consider how you could change this to not run at every boot and respect manual settings

at mytest.py

```
profile("Geolocation")
# Not run at every boot
#import Tools.Geolocation
#Tools.Geolocation.InitGeolocation()
```

at Geolocation.py
```
def InitGeolocation():
	global geolocation
	# Do lookup if not done already, or if info is older than a day
	if "lastchecked" not in geolocation || time.time() - geolocation['lastchecked'] > 86400 :
	try:
		response = urlopen("http://ip-api.com/json/?fields=33288191", data=None, timeout=10).read()
		# print "[Geolocation] DEBUG:", response
		if response:
			geolocation = loads(response)
		status = geolocation.get("status", None)
		if status and status == "success":
			print "[Geolocation] Geolocation data initialised."
			# Timestamp of last check
			geolocation["lastchecked"] = time.time()
		else:
			print "[Geolocation] Error: Geolocation lookup returned a '%s' status!  Message '%s' returned." % (status, geolocation.get("message", None))
	except URLError as err:
		if hasattr(err, 'code'):
			print "[Geolocation] Error: Geolocation data not available! (Code: %s)" % err.code
		if hasattr(err, 'reason'):
			print "[Geolocation] Error: Geolocation data not available! (Reason: %s)" % err.reason
	except ValueError:
		print "[Geolocation] Error: Geolocation data returned can not be processed!"

def RefreshGeolocation():
	# Refresh if needed
	#global geolocation
	#geolocation = {}
	InitGeolocation()
```

Then at Timezone.py
```
def InitTimeZones():
	# if the timezone value not set or set to a invalid one
	if not config.timezone.val.value or timezone_not_valid:
		# Initialize or refresh geolocation
		Tools.Geolocation.RefreshGeolocation()

		tz = geolocation.get("timezone", None)
		proxy = geolocation.get("proxy", False)
		if tz is None or proxy is True:
(...)
```

**Comment by IanSav** [comment]
> I know "?fields=33288191" is static not about me, it fingerprint every enigma2 with your code, to associate with me they have my WAN IP.

No, it is simply a list of requested fields.  *ALL* queries that ask for all fields will use the same number.  This is in no way Enigma2 specific.

Thank you and I appreciate your effort to demonstrate your code proposal.  The code suggestion presumes that the geolocation data is only used for the time zone.  That is currently true as the feature has only just been added but will not hold true in the future.  I am aware of desires to use geolocation data elements for other uses.  I think I heard that it would be helpful for obtaining general location for weather data.  There may well be other uses for the data.

I understand your desire not to use geolocation data so please follow Captain's advice and block the call on your receiver.  In the future when mytest.py is refactored I hope to be able to offer a UI based control to manage the geolocation lookup.

**Comment by Azureit** [comment]
> The code suggestion presumes that the geolocation data is only used for the time zone. That is currently true as the feature has only just been added but will not hold true in the future. I am aware of desires to use geolocation data elements for other uses. I think I heard that it would be helpful for obtaining general location for weather data. There may well be other uses for the data.

I know geolocation will have other uses in future, but any code requesting gelocation.get("WHATEVER") can first call **Tools.Geolocation.RefreshGeolocation()** to initialize or refresh geolocation dict, it makes more sense.
If your code in the future select the language, please respect people choice if they have manually set a different language that your geolocation data is telling you, same thing for the weather app.
I know you code is not finished, but as of now it is not respecting manually configured settings.

**Comment by jose-cruz78** [comment]
wow, this got way out of hand! 

I just want to add that I simply don't understand why so much resistance to make this optional. 
If is complicated to implement, well @IanSav should have thought of that first, no?
'Shoot first and ask questions later'?

I feel users are being left behind to satisfy some contributors personal ideals.

> I am aware of desires to use geolocation data elements for other uses. I think I heard that it would be helpful for obtaining general location for weather data. There may well be other uses for the data.

That brings even more questions, imagine 3rd party plugins accessing that data and then sending it to god knows where.

Weather "data" always worked without the need for that. User manually sets the desired location
(he may want a place other than home)

> follow Captain's advice and block the call on your receiver
Is that all you can give users like me? To be honest, I wouldn't even know how the hell do that!
But don't bother, after my receiver failed to boot and when it worked, it made me thought I was in the US (wrong time), I just reflashed a backup, then someone pointed me to this mess.

**Comment by IanSav** [comment]
@Azureit the solution you suggest to use geolocation calls just before use will initiate even more calls to the geolocation server.  If you have issues with one call I am certain you will have a lot more issues with many repeated calls.

I am not expecting to use the geolocation code to set the default language.  The absolute default locale, by definition is "en_US".  That is how Enigma2 is meant to be written.  All other languages require translation.  That includes "en_GB" which is the general default language offered by most images.  OpenATV defaults to "de_DE".

**Comment by IanSav** [comment]
@jose-cruz78 why do you believe this is out of hand?

These changes were specifically made to enhance the user experience and to allow Enigma2 to make more user friendly initial choices for you.  Apparently there are users who do not appear to be happy to configure some aspects of their receivers.  This code tries to make better *initial* guesses for them.  Once the initial guess is made the geolocation data is no longer used.  (As I have said a number of times now, the current implementation of Enigma2 does not allow sufficient control at the very early stages of Enigma2 startup.  When I, or another developer, changes that then users will be able to have much more control on how their receivers start up.)

You and others seem to think that I am leaking private information out of your receiver.  That is *not* the case.  All this geolocaction call does is collect *public* information that is already available about your IP address for *your* use.  *NOTHING* is sent from your receiver other than a completely anonymous request to have *publicly available* information about your WAN connection returned to you.

I bet the few of you raising concerns continue to use the web, social media and other Internet based activities.  How you directed these sort of questions and accusations to all those providers?  Believe it or not most people on the Internet are performing significantly more intensive geolocation calls on your behalf.  They are combining that with other data, often provided by you, to do significantly more than offer a suggested time zone for your receiver to use.  (Note that on Enigma2 this geolocation code does an anonymous geolocation call and collects the results itself.  The results are not kept or used for *any* other purpose that to initialise some functions on your receiver.)

Next time you surf the web with your private browsing options turned on, your cookies turned off and the connection set to its most private settings please tell why the website you ace connecting to knows to present the page with your country's flag and your local time.  It may even use a language of your country even if your browser's language is different.  The reason is that the website you are using is doing a 3rd party geolocation lookup of *your* connection.  Are you complaining to all those websites about invasion of your privacy?

The only reason this thread is getting out of hand is that some users want to believe that I have in some unspecified way taken their *private* information and leaked it to the Internet.  That is simply *NOT* true.  Perhaps I am encouraging the "mess" by being open and communicative and trying the explain exactly what is going on.  (Apparently to little avail as people simply don't want to listen or understand me.  Everything I am saying is fully verifiable by looking at the Enigma2 open source code, reading the privacy statement of the geolocation provider I am using  and researching geolocation on the Internet.)

**Comment by jose-cruz78** [comment]
@IanSav not be foolish, you are repeting yourself over and over, without really understanding the users base. What other 'websites' do or not is not our concern here. We are talking about this project not other web. Also, no one makes me open some web page, but you make people use this.
Maybe you trying to make enigma idiot proof and forgeting others.
Even people like me (can) limit and restrict private data shared online to a certain level using vpns or even tor. 
I do not want enter in meaninless discussions with you, many above already.

+1 for making this optional, if not, better remove it.

**Comment by IanSav** [comment]
None so deaf as those that will not hear. None so blind as those that will not see.

**Comment by IanSav** [comment]
@jose-cruz78 where is your receiver getting its EPG?

**Comment by jose-cruz78** [comment]
Internet yes, no epg from internet.

**Comment by IanSav** [comment]
So have you complained to all the Internet connections providers you have made that did 3rd party geolocation lookups on you?  Trust me that they did look you up.  They didn't tell you, they didn't get your consent.  They didn't give you an option to opt out.

**Comment by IanSav** [comment]
> Internet yes, no epg from internet.
> I told you, I was pointed here, didn't know github before.

You created an account and gave GitHub (owned by Microsoft) far more information than my geolocation code returned to you!

**Comment by Azureit** [comment]
Let's all calm down please, inflamed commend don't lead us nowhere.

@IanSav 
> @Azureit the solution you suggest to use geolocation calls just before use could initiate even more calls to the geolocation server. If you have issues with one call I am certain you will have a lot more issues with many repeated calls.

Look at the patch at my comment, I even comment it:
```
	# Do lookup if not done already, or if info is older than a day
	if "lastchecked" not in geolocation || time.time() - geolocation['lastchecked'] > 86400 :
```

**Comment by IanSav** [comment]
People need to get a sense of proportion here.  Geolocation is a fact of life.  The code I used is clear and open.  Nothing is hidden.  No private information is obtained or provided to the geolocation provider.  The information that is returned is completely public and readily available.

If I have not clearly documented the addition of the geolocation tool I suggest that none of this discussion would occur.  People would never have noticed notices it.  I am being very clear and open about what is going on.

**Comment by jose-cruz78** [comment]
> So have you complained to all the Internet connections providers you have made that did 3rd party geolocation lookups on you? Trust me that they did look you up. They didn't tell you, they didn't get your consent. They didn't give you an option to opt out.
You created an account and gave GitHub (owned by Microsoft) far more information than my geolocation code returned to you!

It seems you are desperate defending your thing
You tell me like, others do it like this, so is ok here?
Is ok that I kill somenone?  ...others are doing it!

**Comment by IanSav** [comment]
@Azureit the data can change with every connection.  The only way to validate the cache would be to perform another lookup.  That defeats any gain by caching the data.  The opposite is true, the code to do all the work simply becomes bloat.

**Comment by IanSav** [comment]
@jose-cruz78 I have nothing to defend.  I am simply being polite and responding to address your concerns.  By the tone and nonsense of your last post I will take it that you wish me to ignore your posts.

**Comment by Azureit** [comment]
> @Azureit the data can change with every connection. The only way to validate the cache would be to perform another lookup. That defeats any gain by caching the data. The opposite is true, the code to do all the work simply becomes bloat.

It's not caching we do the request to populate the geolocation dictionary and add an extra key "lastchecked" with the current timestamp and then next time geolocation is needed we check the key to know we already have fresh geolocation and skip getting new.

**Comment by IanSav** [comment]
What you propose is the definition of caching!

**Comment by IanSav** [comment]
Three people have now negatively commented on the geolocation change.  Why can't any of you demonstrate how privacy is being compromised by this code?

**Comment by Azureit** [comment]
@IanSav  but you do caching of geolocation dictionary already..

**Comment by IanSav** [comment]
> @IanSav but you do caching of geolocation dictionary already..

Yes, once a lookup is done I don't do it again until the box restarts.  I felt that was appropriate level of caching.  I provide the ability to refresh the cache should an application desire the latest update.

**Comment by Azureit** [comment]
@IanSav 
> Yes, once a lookup is done I don't do it again until the box restarts. I felt that was an appropriate level of caching. I provide the ability to refresh the cache more frequently should an application desire the latest update.

My way would not require lookup at boot, and would provide refresh if data older than a day, but you can tweak that down.

>

**Comment by IanSav** [comment]
The geolocation call is a small data request with minimal data exchange.  What exactly is your issue with the lookup?

**Comment by Azureit** [comment]
> The geolocation call is a small data request with minimal data exchange. What exactly is your issue with the lookup?

Why would you do a request if needed for nothing, all manually configured.
My enigma box turn off with my TV.

**Comment by IanSav** [comment]
As I updated my previous post.  The design and implementation of Enigma2 is currently flawed such that user configuration settings are only enabled at the *end* of the system startup.

**Comment by Azureit** [comment]
> As I updated my previous post. The design and implementation of Enigma2 is currently flawed such that user configuration settings are only enabled at the end of the system startup.

That's the problem, you should put user settings first.

> Refactoring mytest.py is a significant effort. It is not trivial to make fixes in this code.
In my view you just comment the **Tools.Geolocation.InitGeolocation()** and do that before geolocation.get("WHATEVER")

**Comment by Azureit** [comment]
@IanSav 
> If I refactor this code I intend to make loading the user's preferences and settings the first thing the code does. That will enable me to skip the geolocation call completely if the user has already configured their preferred settings. This will also allow a significantly more optimised start up experience.

That would be great, thanks.

**Comment by IanSav** [comment]
That is just adding more bad code that will have to be fixed again later.

Please tell me exactly what is the problem with the geolocation lookup?  I am not talking about the code I am asking about the core reason for the objection to the concept within Enigma2.  This issue has still not been made clear.  I have, I believe, addressed every privacy concern raised so far.

**Comment by Azureit** [comment]
@IanSav  Read my post above, this is what I would like to read from you that you intend to load user settings first and skip the geolocation if not need. Case closed for me.
Thanks

**Comment by IanSav** [comment]
> That would be great, thanks.

I actually tried to make this change when the first concern was raised.  Trust me when I say the Enigma2 crashed and burned.  So many people have tried to work around this design flaw that when I moved the code *many* things broke.  :(

I am definitely not ignoring the concerns expressed here it is just that fixing them is *VERY* complicated and will have widespread implications to much of Enigma2.

**Comment by IanSav** [comment]
You can also trust me that if I do try the refactor mytest.py I will receive many more complaints and objections.  That change will be *VERY* hard to get accepted.  A lot of Enigma2 developers resist and/or hate change.

(You may not be aware but I am a developer for OpenATV, OpenPLi and OpenViX.  My code is also used in a number of other derivative Enigma2 images.  I try to develop the same code and changes for all the images.  It is one of my objectives to improve the user interface of Enigma2 for all images and to try and make more code, plugins and skins interoperable between the images.  The same geolocation code is now running in at least those main images and may already have been integrated into the derivative images.)

**Comment by Azureit** [comment]
> You can also trust me that if I do try the refactor mytest.py I will receive many more complaints and objections. That change will be VERY hard to get accepted. A lot of Enigma2 developers resist and/or hate change.

I'm starting to worry about that "if", please the right way to do it, is to load user settings first and skip if not needed.
> You may not be aware but I am a developer for OpenATV, OpenPLi and OpenViX. My code is also used in a number of other derivative Enigma2 images

That's because I tell you "?fields=33288191" is a way to fingerprint the use of an enigma2, trust me most of the requests for other developers don't need the fields you need.

**Comment by IanSav** [comment]
> That's because I tell you "?fields=33288191" is a way to fingerprint the use of an enigma2, trust me most of the requests for other developers don't need the fields you need.

We have been through this.  It is NOT a fingerprint.  Any request to get the full data set will use the same number.  A representative of the company has already confirmed that the data is not a fingerprint and is not kept or used in any way other to return the data requested.

**Comment by IanSav** [comment]
It is now 7AM here (Australia) I have spent most of the night responding to the concerns expressed here.  I trust that you will see that I am taking your comments seriously.  Unfortunately I am now way over tired and am going to get some sleep.

**Comment by jose-cruz78** [comment]
> > You can also trust me that if I do try the refactor mytest.py I will receive many more complaints and objections. **That change will be VERY hard to get accepted**. A lot of Enigma2 developers resist and/or hate change.
> 
> I'm starting to worry about that "if", please the right way to do it, is to load user settings first and skip if not needed.

Worry more because it could also mean **never**!

**Comment by IanSav** [comment]
@jose-cruz78 you do seem to be a glass half empty kind of person.  :(

**Comment by Tony-il-Capo** [comment]
I started to build OpenATV 6.4 with all geolocation code reverted. Local demand is high and I can't handle build for different hardware. Build machine is not very fast! 😓 
Wacky workarounds won't do..., so I started to share old python files (mystest.py(o), SetupDevices.pyo, Timezones.pyo,..) so people can revert to old behavior more easily.
Also, kinda _wacky_ too, and of course, this will break if someone tries to update, but it effectively removes it.

I kindly ask @IanSav to at least consider reverting geolocation changes in the '6.3' branch. This approach could also be taken to start refactoring mystest.py code, could even move all geolocation code there ( to the new testing brach?), so, when it goes 'public', it already gives the user the power to control it using UI. Everyone's happy!
At the moment, there's no alternative downloads at opena.tv, unless people go for a very outdated image. (6.2-)

Thanks for your time
Bye.

**Comment by IanSav** [comment]
@Tony-il-Capo I only contribute to the 6.4 branch (as requested by Captain).  I am not involved with code being moved to the 6.3 branch.

If you start modifying and diverting OpenATV to sabotage the new code then I suggest that you confirm this course of action with the OpenATV image administrators.  This action is in direct contradiction to their requests that I include OpenATV in the Enigma2 improvement and common code developments.  I will not be allocating limited resources to fixing OpenATV and undoing damage to future merges that your changes are likely to create!  Your refusal to accept that geolocation is totally private and harmless is your issue.  Changing OpenATV to interfere with the development efforts is going to hurt many others.

By the way, I have already started initial development work to refactor mytest.py to resolve the issues previously discussed.

**Comment by Azureit** [comment]
@jose-cruz78 
> Worry more because it could also mean **never**!

This is a open source project, and no one pays for it, so there is no accountability to the end user. 
@IanSav don't share my privacy concerns but I respect him for keeping an open dialogue, most developers at his position would just say no and ignore you.

**Comment by Azureit** [comment]
@IanSav
Lets solve this in steps.
step 1: auto initialization/auto refresh geolocation data
step 2: read manual settings first
I will forget for now step 2 and concentrate at step 1.

My vision was that no initialization should be needed, a simple geolocation.get("WHATEVER") should initialize the geolocation dictionary. 
Disclaimer: My knowledge about python is by example, nevertheless I research this and arrive to what maybe a solution for this, that is working and tested by me, with minimal changes to your code.

```
--- <mytest.py>
+++ <mytest.py>
@@ -29,8 +29,7 @@
 from traceback import print_exc
 
 profile("Geolocation")
-import Tools.Geolocation
-Tools.Geolocation.InitGeolocation()
+# lookup when needed
 
 profile("SetupDevices")
 import Components.SetupDevices

```

```
--- <Geolocation.py>
+++ <Geolocation.py>
@@ -1,7 +1,8 @@
 from json import loads
 from urllib2 import URLError, urlopen
+from time import time
 
-# Data available from http://ip-api.com/json/:
+# Data available from http://ip-api.com/json/?fields=33288191:
 #
 # 	Name		Description				Example			Type
 # 	--------------	--------------------------------------	----------------------	------
@@ -37,30 +38,31 @@
 # 	hosting		Hosting, colocated or data center	true			bool
 # 	query		IP used for the query			173.194.67.94		string
 
-geolocation = {}
+class Geolocation(dict):
+	def get(self, k, default=None):
+		# check interval = 1 day
+		if 'lastchecked' not in self or time() - self['lastchecked'] > 86400:
+				try:
+					response = urlopen('http://ip-api.com/json/?fields=33288191', data=None, timeout=10).read()
+					# print "[Geolocation] DEBUG:", response
+					if response:
+						try:
+							self.update(loads(response))
+						except:
+							self = {}
+					status = super(Geolocation, self).get('status', None)
+					if status and self['status'] == 'success':
+						print "[Geolocation] Geolocation data initialised."
+						self['lastchecked'] = int(time())
+					else:
+						print "[Geolocation] Error: Geolocation lookup returned a '%s' status!  Message '%s' returned." % (status, super(Geolocation, self).get("message", None))
+				except URLError as err:
+					if hasattr(err, 'code'):
+						print "[Geolocation] Error: Geolocation data not available! (Code: %s)" % err.code
+					if hasattr(err, 'reason'):
+						print "[Geolocation] Error: Geolocation data not available! (Reason: %s)" % err.reason
+				except ValueError:
+					print "[Geolocation] Error: Geolocation data returned can not be processed!"
+		return super(Geolocation, self).get(k, default)
 
-def InitGeolocation():
-	global geolocation
-	if len(geolocation) == 0:
-		try:
-			response = urlopen("http://ip-api.com/json/?fields=33288191", data=None, timeout=10).read()
-			# print "[Geolocation] DEBUG:", response
-			if response:
-				geolocation = loads(response)
-			status = geolocation.get("status", None)
-			if status and status == "success":
-				print "[Geolocation] Geolocation data initialised."
-			else:
-				print "[Geolocation] Error: Geolocation lookup returned a '%s' status!  Message '%s' returned." % (status, geolocation.get("message", None))
-		except URLError as err:
-			if hasattr(err, 'code'):
-				print "[Geolocation] Error: Geolocation data not available! (Code: %s)" % err.code
-			if hasattr(err, 'reason'):
-				print "[Geolocation] Error: Geolocation data not available! (Reason: %s)" % err.reason
-		except ValueError:
-			print "[Geolocation] Error: Geolocation data returned can not be processed!"
-
-def RefreshGeolocation():
-	global geolocation
-	geolocation = {}
-	InitGeolocation()
+geolocation = Geolocation()

```

**Comment by IanSav** [comment]
@Azureit it turns out that other developers are also annoyed about the initialisation issues in Enigma2.  I am now working with another OpenViX developer to refactor, and possibly rename, mytest.py to fix these problems once and for all time.  When we get the code working on OpenViX I will work with the OpenATV developers to try and make equivalent corrections to OpenATV.  (Not that it matters here but the same will also happen with OpenPLi.)

I have already PEP8 cleaned the code and added lots of debugging code to work out what is going on now.  I am getting some sleep now but tomorrow when Tony (OpenViX) and I are both available we are going to give this proposal a really good try.

**Comment by Azureit** [comment]
@IanSav Thanks for using your time to help the enigma2 community.

**Comment by Azureit** [comment]
@Tony-il-Capo 
> I started to build OpenATV 6.4 with all geolocation code reverted. Local demand is high and I can't handle build for different hardware. Build machine is not very fast! sweat
Wacky workarounds won't do..., so I started to share old python files (mystest.py(o), SetupDevices.pyo, Timezones.pyo,..) so people can revert to old behavior more easily.
Also, kinda wacky too, and of course, this will break if someone tries to update, but it effectively removes it.

**I didn't test this and may not work or break your box**, but if wanted to disable geolocation without recompiling, I would target the rootfs.bin update file. Edit with an hexeditor and search for the string "**ip-api.com**" and replace with a string of same length like "**127.0.0.10**"
### I didn't test this and may not work or break your box

**Comment by IanSav** [comment]
> @IanSav Thanks for using your time to help the enigma2 community.

This is a major project and change for Enigma2.  I appreciate your acknowledgement but this will be a significant effort from a number of developers across a number of images.

**Comment by Tony-il-Capo** [comment]
> Worry more because it could also mean never!

It was like a :rocket: to implement... but now a :turtle: to fix
@jose-cruz78 
I've seen evidence of fingerprinting... it's probably not so hard for (google) algorithms to pick up this code and scan other open source for the same fields requested. Probably not many or even nothing more asks for so many fields.
Why the hell, lat, lon, ISP, org, currency.. is needed?

We have ISPs monitoring every connection made, and share data with third parties (ie google), here a guy that has like 7 stbs, shutting down/starting 3-4 times a day, found himself flooded with E2 related ads when browsing. The guy only browses email and weather. wtf?!

If  'some' could really see how many people reverted/stopped updating their images... but, I guess, someone with limited perspective, that judges their 'awesome' commits to E2 by the reaction of their dev pals and ignore real user base, can't really see that.

**Comment by jenny-91** [comment]
Hey
I don't use OpenATV anymore, but recently noticed this geolocation cr*p code in other dists, I searched and found this thread and decided to sign in and share my opinon.
Afterwards I removed/disabled it for my builds. Luckly, ip-api.com is listed as a tracker in any tracking lists and my network setup blocked it anyway.

I have to say I'm very disapointed in what E2 project is becoming, thoughtless implementations will destroy this project. Contributors that don't value user opinions and refuse to understand them, should stay away from projects like this.

**Comment by IanSav** [comment]
@Jenny-91, can you please provide *proof* that the geolocation provider is tracking users?

The *only* data they get from Enigma2 users is their publically available WAN IP address.  I have checked their privacy policy and spoken with the developers/maintainers of that site and received assurance that there is no use of any connection data, other than to return geolocation data and there is no long term retention of data other than the short term retention (in minutes) to detect service abuse.

Given your feeling for geolocation I trust that you don't use the Internet.  If you do your are most certainly using geolocation services.  The main difference is that here that data is being returned to you.  On the vast majority of Internet servers the geolocation data is being given to the website provider.

I should also reiterate that I am working on changing mytest.py to allow users to control if the geolocation, and other initialisation tasks, are run at all.

**Comment by AbuBaniaz** [comment]
@jenny-91 
Can you share how to disable these checks? 

It does not even work.
So every time the receiver reboots, it is checking/sharing your location without your permission. And does not even fulfil its purpose. If You flash PLI/ATV in the UK. Your timezones start as Berlin or Amsterdam respectively. I am sure if you flash ViX In the Netherlands, you will start off as London. 

The website being used says it passes on some IP ranges to another website.  There is no mention of which ranges. 

In the second company's policy they state.

Information Collected Automatically
We automatically collect information about your device and how your device interacts with our Services and other services. The categories of information we automatically collect, including in the last 12 months, and the methods by which we may collect such information include the following:
(I have not coped rest of text, find it yourself)

Location Data
MaxMind may also request access to or otherwise receive information about your device location when you access the Services. Your location data may be based on your IP address and other location-aware technologies. We use location data in connection with providing the Services and to help improve the Services.

**Comment by atvcaptain** [comment]
https://github.com/openatv/enigma2/commit/a35571adddd18a22c95b006fce7d56d669794a53
